### PR TITLE
Fix clz

### PR DESCRIPTION
**Pull request by owolabileg**
This PR fixes `-DdepFormat=ZLC` that was previously broken.

@august782 Please review

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,4 +1,12 @@
...
+os:
+    - linux
+
 before_install:
-    - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then jdk_switcher use oraclejdk8; fi
\ No newline at end of file
+    - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then jdk_switcher use oraclejdk8; fi
+
+env:
+    - CMD="mvn clean verify -B"
```
Comment (august782) [inline .travis.yml]: Why not just use "mvn verify -B" under script?

File: starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java [hunk 2]
```diff
@@ -39,41 +45,48 @@
...
-        for (String line : Arrays.asList(baosErr.toString().split("\n"))) {
-            if (line.contains("::Diff::")) {
-                // This assumes that only classes in the application can
-                // change, and not classes in jars
-                String fqn = Writer.toFQN(line);
-                changed.add(fqn);
+        for (String line : Arrays.asList(baosErr.toString().split(lineSeparator))) {
+            String ekstaziDiffMarker = "::Diff:: ";
+            if (line.contains(ekstaziDiffMarker)) {
+                changed.add(line.split(ekstaziDiffMarker)[1]);
```
Comment (august782) [inline starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java]: Before changed was adding the toFQN of the line, while this new one takes the second half of ekstaziDiffMarker. Does this lead to a difference?

File: starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java [hunk 3]
```diff
@@ -99,8 +96,8 @@ protected void updateForNextRun(Set<String> nonAffected) throws MojoExecutionExc
         Set<String> affectedTests = new HashSet<>(allTests);
         affectedTests.removeAll(nonAffected);
         DirectedGraph<String> graph = null;
-        Map<String, Set<String>> testDeps = null;
-        Set<String> unreached = null;
+        Map<String, Set<String>> testDeps;
```
Comment (august782) [inline starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java]: Is it necessary to even define these variables here as they are not used in this scope afterwards?

File: starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java [hunk 4]
```diff
@@ -39,41 +45,48 @@
...
+    public static Pair<Set<String>, Set<String>> getNonAffectedTests(File basedir) {
         long start = System.currentTimeMillis();
-        List<String> nonAffected = AffectedChecker.findNonAffectedClasses(basedir, getRootDirOption(basedir));
+        List<String> nonAffectedFiles = AffectedChecker.findNonAffectedClasses(basedir, getRootDirOption(basedir));
         long end = System.currentTimeMillis();
+        Set<String> nonAffectedTests = toFQN(new HashSet<>(nonAffectedFiles));
+        Set<String> changed = new HashSet<>();
         LOGGER.log(Level.FINEST, "[TIME]COMPUTING NON-AFFECTED(2): " + (end - start) + "ms");
-        return toFQN(new HashSet<>(nonAffected));
+        return new Pair<>(nonAffectedTests, changed);
```
Comment (august782) [inline starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java]: To clarify, changed will always be empty, right?

File: starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java [hunk 5]
```diff
@@ -69,25 +70,21 @@ public void execute() throws MojoExecutionException {
...
-            //TODO: parse .clz files to find what changed in the same pass as finding nonaffected tests
-            Set<String> changed = new HashSet<>();
-            //TODO: updateForNextRun, depending on updateChecksums
-            return new Pair<>(nonAffected, changed);
+            data = EkstaziHelper.getNonAffectedTests(getArtifactsDir());
         }
-        return null;
+        Set<String> changed = data == null ? new HashSet<String>() : data.getValue();
+        if (Logger.getGlobal().getLoggingLevel().intValue() <= Level.FINEST.intValue()) {
+            Writer.writeToFile(changed, "changed-classes", getArtifactsDir());
```
Comment (august782) [inline starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java]: In the case of CLZ, there would be empty changed, and so this just prints nothing, right?

File: starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java [hunk 6]
```diff
@@ -69,25 +70,21 @@ public void execute() throws MojoExecutionException {
...
-            Pair<Set<String>, Set<String>> data = zlcHelper.getChangedData(getArtifactsDir(), cleanBytes);
-            Set<String> changed = data == null ? new HashSet<String>() : data.getValue();
-            if (Logger.getGlobal().getLoggingLevel().intValue() <= Level.FINEST.intValue()) {
-                Writer.writeToFile(changed, "changed-classes", getArtifactsDir());
-            }
-            long end = System.currentTimeMillis();
-            Logger.getGlobal().log(Level.FINE, "[PROFILE] COMPUTING CHANGES: " + Writer.millsToSeconds(end - start));
-            return data;
+            data = zlcHelper.getChangedData(getArtifactsDir(), cleanBytes);
         } else if (depFormat == DependencyFormat.CLZ) {
```
Comment (august782) [inline starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java]: There are only two formats, ZLC and CLZ, right? If some other format makes it through (not sure if it's possible), the current logic just results in some default, empty returns, right? Should an exception be used instead, indicating some big break?

File: starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java (hunk 4 above)
Comment (owolabileg) [reply starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java]: In this overloaded method, yes. Calling `AffectedChecker.findNonAffectedClasses()` does not return the changed classes.

File: starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java (hunk 2 above)
Comment (owolabileg) [reply starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java]: Yes, we now store the `URL.toExternalForm()` to be consistent with what we store with `-DdepFormat=ZLC`

File: starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java (hunk 6 above)
Comment (owolabileg) [reply starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java]: Currently, if you run with a format that is not in `{CLZ,ZLC}`, e.g., `TTT`, you get an `IllegalArgumentException` exception: `Cannot convert 'TTT' to Enum: No enum constant edu.illinois.starts.enums.DependencyFormat.TTT`; that `depFormat` argument is backed by an `enum`

File: starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java (hunk 5 above)
Comment (owolabileg) [reply starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java]: There are two overloaded methods in `EkstaziHelper`; the one we are using does return a non-empty `changed` if run in the right logging level and the ekstazi config (`.ekstazirc`) is set correctly.

File: starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java (hunk 3 above)
Comment (owolabileg) [reply starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java]: I fixed that. Thanks!

File: starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java [hunk 7]
```diff
@@ -39,41 +45,48 @@
...
+    }
+
+    private static boolean isFirstRun(String artifactsDir) {
+        // If the notFirstRunMarker file does not exist, this is a first run
+        return !(new File(artifactsDir, notFirstRunMarker).exists());
     }
 
-    private static void writeEkstaziDebugInfo(ByteArrayOutputStream baosErr, String artifactsDir) {
+    private static Set<String> writeEkstaziDebugInfo(ByteArrayOutputStream baosErr, String artifactsDir) {
```
Comment (august782) [inline starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java]: Name is a bit non-indicative of it returning changed classes. Maybe either refactor the name of the method or add a quick comment with the method as a reminder.

File: starts-core/src/main/java/edu/illinois/starts/helpers/Writer.java [hunk 8]
```diff
@@ -172,6 +172,9 @@ public static String toFQNPlain(String line) {
     public static List<String> fqnsToExcludePath(Collection<String> fqns) {
         List<String> paths = new ArrayList<>();
         for (String fqn : fqns ) {
```
Comment (august782) [inline starts-core/src/main/java/edu/illinois/starts/helpers/Writer.java]: Cut whitespace after fqns

File: starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java [hunk 9]
```diff
@@ -69,25 +70,21 @@ public void execute() throws MojoExecutionException {
...
-            Set<String> nonAffected = EkstaziHelper.getNonAffectedTests(getArtifactsDir());
-            //TODO: parse .clz files to find what changed in the same pass as finding nonaffected tests
-            Set<String> changed = new HashSet<>();
-            //TODO: updateForNextRun, depending on updateChecksums
-            return new Pair<>(nonAffected, changed);
+            data = EkstaziHelper.getNonAffectedTests(getArtifactsDir());
```
Comment (august782) [inline starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java]: For CLZ, it is unnecessary to pass around the cleanBytes flag as an option?

File: starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java (hunk 7 above)
Comment (owolabileg) [reply starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java]: changed name and added comment

File: starts-core/src/main/java/edu/illinois/starts/helpers/Writer.java (hunk 8 above)
Comment (owolabileg) [reply starts-core/src/main/java/edu/illinois/starts/helpers/Writer.java]: Done. I also added `<module name="ParenPad"/>` to `checkstyle.xml` so we should catch this in the future.

File: starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java (hunk 9 above)
Comment (owolabileg) [reply starts-plugin/src/main/java/edu/illinois/starts/jdeps/DiffMojo.java]: Unfortunately, yes. AFAIK, we have to control the `cleanBytes` for Ekstazi via the `hash.without.debuginfo = true` setting in the `.ekstazirc` file.

**Comment by owolabileg** [comment]
@august782 Let me know if you have more comments on this PR, or whether I should merge now.

File: starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java [hunk 10]
```diff
@@ -39,41 +45,56 @@
...
     }
 
-    private static void writeEkstaziDebugInfo(ByteArrayOutputStream baosErr, String artifactsDir) {
+    /**
+     * Process the Ekstazi debug output to get what classes Ekstazi thinks changed, and write those changed classes
+     * to file.
+     *
+     * @param baosErr      Ekstazi Debug Output
+     * @param artifactsDir Directory in which we store STARTS artifacts (i.e., ".starts")
+     * @return             The (possibly empty) list of changed classes
```
Comment (august782) [inline starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java]: It's actually a set

Comment (owolabileg) [reply starts-core/src/main/java/edu/illinois/starts/helpers/EkstaziHelper.java]: fixed

**Pull request merged**